File: src/v3/runtime/replay.test.ts
import { describe, expect, it } from 'vitest';
import { isDiceRollError } from '../errors.js';
import { createExecutionContext } from './context.js';
import {
  canonicalizeSeed,
  createAutomaticSeed,
  createProvidedSeed,
  createReplayDescriptor,
  createReplaySeed,
  isReplayDescriptor,
  validateReplayDescriptor,
  type CryptoSource,
  type ReplayDescriptor,
} from './replay.js';

const PLAN_FINGERPRINT = '0123456789abcdef0123456789abcdef';

function descriptor(overrides: Partial<ReplayDescriptor> = {}): ReplayDescriptor {
  return {
    schemaVersion: 2,
    algorithm: 'mt19937',
    algorithmVersion: 1,
    executionVersion: 1,
    mathProfile: 'decimal12-v1',
    origin: 'provided-string',
    seedMaterial: '0123456789abcdef0123456789abcdef',
    planFingerprint: PLAN_FINGERPRINT,
    ...overrides,
  };
}

describe('seed and replay V2', () => {
  it('canonicalizes number and string seeds into separate fixed materials', () => {
    expect(canonicalizeSeed(42)).toBe('number:42');
    expect(canonicalizeSeed('42')).toBe('string:42');
    expect(createProvidedSeed(42).origin).toBe('provided-number');
    expect(createProvidedSeed('42').origin).toBe('provided-string');
    expect(createProvidedSeed(42).seedMaterial).toMatch(/^[0-9a-f]{32}$/u);
    expect(createProvidedSeed(42).words).not.toEqual(createProvidedSeed('42').words);
    expect(canonicalizeSeed(-0)).toBe('number:-0');
  });

  it('enforces text seed length before hashing', () => {
    expect(canonicalizeSeed('abcd', 4)).toBe('string:abcd');
    expect(() => canonicalizeSeed('abcde', 4)).toThrow(
      expect.objectContaining({ code: 'INVALID_SEED' }),
    );
    expect(() => canonicalizeSeed('x', 0)).toThrow(
      expect.objectContaining({ code: 'INVALID_LIMIT' }),
    );
  });

  it('creates a 128-bit automatic seed from crypto without exposing source text', () => {
    const cryptoSource: CryptoSource = {
      getRandomValues(values: Uint32Array): Uint32Array {
        values.set([1, 2, 3, 4]);
        return values;
      },
    };
    const seed = createAutomaticSeed(cryptoSource);
    const replay = createReplayDescriptor(seed, { planFingerprint: PLAN_FINGERPRINT });

    expect(seed.seedMaterial).toBe('00000001000000020000000300000004');
    expect(replay).toEqual({
      schemaVersion: 2,
      algorithm: 'mt19937',
      algorithmVersion: 1,
      executionVersion: 1,
      mathProfile: 'decimal12-v1',
      origin: 'crypto',
      seedMaterial: seed.seedMaterial,
      planFingerprint: PLAN_FINGERPRINT,
    });
    expect(JSON.stringify(replay)).not.toContain('crypto:');
  });

  it('fails without working crypto and rejects non-finite numeric seeds', () => {
    expect(() => createAutomaticSeed(null)).toThrow(
      expect.objectContaining({ code: 'RNG_UNAVAILABLE' }),
    );
    expect(() => createAutomaticSeed({
      getRandomValues(): Uint32Array {
        throw new Error('unavailable');
      },
    })).toThrow(expect.objectContaining({ code: 'RNG_UNAVAILABLE' }));
    expect(() => canonicalizeSeed(Number.POSITIVE_INFINITY)).toThrow(
      expect.objectContaining({ code: 'INVALID_SEED' }),
    );
  });

  it('replays both algorithms exactly and binds descriptors to a plan', () => {
    for (const algorithm of ['mt19937', 'xoshiro128ss'] as const) {
      const first = createExecutionContext({
        seed: 'encounter-42',
        randomAlgorithm: algorithm,
        planFingerprint: PLAN_FINGERPRINT,
      });
      const second = createExecutionContext({
        replay: first.replay,
        planFingerprint: PLAN_FINGERPRINT,
      });
      expect(first.replay).toEqual(second.replay);
      expect(Array.from({ length: 20 }, () => first.random.integer(1, 100))).toEqual(
        Array.from({ length: 20 }, () => second.random.integer(1, 100)),
      );
    }

    expect(() => createReplaySeed(descriptor(), 'ffffffffffffffffffffffffffffffff'))
      .toThrow(expect.objectContaining({ code: 'REPLAY_PLAN_MISMATCH' }));
  });

  it('validates the complete descriptor including null, extra fields and overrides', () => {
    expect(isReplayDescriptor(descriptor())).toBe(true);
    expect(isReplayDescriptor(null)).toBe(false);
    expect(isReplayDescriptor({ ...descriptor(), extra: true })).toBe(false);
    expect(() => validateReplayDescriptor(null)).toThrow(
      expect.objectContaining({ code: 'INVALID_REPLAY' }),
    );
    expect(() => validateReplayDescriptor({ ...descriptor(), schemaVersion: 1 }))
      .toThrow(expect.objectContaining({ code: 'UNSUPPORTED_REPLAY_VERSION' }));
    expect(() => validateReplayDescriptor({ ...descriptor(), seedMaterial: 'nope' }))
      .toThrow(expect.objectContaining({ code: 'INVALID_REPLAY' }));
    expect(() => createExecutionContext({
      replay: descriptor(),
      randomAlgorithm: 'mt19937',
    })).toThrow(expect.objectContaining({ code: 'INVALID_REPLAY' }));
    expect(() => createExecutionContext({
      seed: 'x',
      replay: descriptor(),
    })).toThrow(expect.objectContaining({ code: 'INVALID_REPLAY' }));
  });

  it('rejects hostile getters without leaking their exception', () => {
    const hostile = Object.defineProperty({}, 'schemaVersion', {
      enumerable: true,
      get(): never {
        throw new Error('hostile');
      },
    });
    try {
      validateReplayDescriptor(hostile);
    } catch (error: unknown) {
      expect(isDiceRollError(error)).toBe(true);
    }
    const revoked = Proxy.revocable({}, {});
    revoked.revoke();
    expect(() => validateReplayDescriptor(revoked.proxy)).toThrow(
      expect.objectContaining({ code: 'INVALID_REPLAY' }),
    );
  });
});

File: go/replay_test.go
package dicecore

import (
	"encoding/json"
	"reflect"
	"testing"
)

const testReplayFingerprint = "0123456789abcdef0123456789abcdef"

func testReplayDescriptor() ReplayDescriptor {
	return ReplayDescriptor{
		SchemaVersion: 2, Algorithm: MT19937, AlgorithmVersion: 1, ExecutionVersion: 1,
		MathProfile: "decimal12-v1", Origin: SeedProvidedString,
		SeedMaterial: "0123456789abcdef0123456789abcdef", PlanFingerprint: testReplayFingerprint,
	}
}

func TestReplayRestoresBothAlgorithms(t *testing.T) {
	seed, _ := CreateProvidedSeed("encounter-42")
	for _, algorithm := range []RandomAlgorithm{MT19937, Xoshiro128SS} {
		replay, err := CreateReplayDescriptor(seed, ReplayDescriptorOptions{Algorithm: algorithm, PlanFingerprint: testReplayFingerprint})
		if err != nil {
			t.Fatal(err)
		}
		state, err := CreateReplayState(replay, testReplayFingerprint)
		if err != nil || state.Replay != replay || state.Seed.Words != seed.Words || state.Seed.Origin != seed.Origin {
			t.Fatalf("replay state: %+v, %v", state, err)
		}
		first, err := NewReplayRandom(replay, nil, testReplayFingerprint)
		if err != nil {
			t.Fatal(err)
		}
		var second RandomSource
		if algorithm == MT19937 {
			second, _ = NewMersenneTwister19937FromWords(seed.Words[:], nil)
		} else {
			second, _ = NewXoshiro128StarStar(seed.Words[:], nil)
		}
		if !reflect.DeepEqual(randomValues(t, first, 700), randomValues(t, second, 700)) {
			t.Fatalf("restored %s sequence diverges", algorithm)
		}
		encoded, _ := json.Marshal(replay)
		var decoded ReplayDescriptor
		if err := json.Unmarshal(encoded, &decoded); err != nil || decoded != replay {
			t.Fatalf("JSON roundtrip: %+v, %v", decoded, err)
		}
	}
}

func TestReplayStrictValidation(t *testing.T) {
	valid := testReplayDescriptor()
	if !IsReplayDescriptor(valid) || !IsReplayDescriptor(&valid) {
		t.Fatal("valid descriptor rejected")
	}
	_, err := ValidateReplayDescriptor(valid, "ffffffffffffffffffffffffffffffff")
	assertRandomErrorCode(t, err, "REPLAY_PLAN_MISMATCH")
	for _, value := range []any{nil, []any{}, "invalid", (*ReplayDescriptor)(nil), map[string]any(nil)} {
		_, err := ValidateReplayDescriptor(value)
		assertRandomErrorCode(t, err, "INVALID_REPLAY")
	}
	for _, mutate := range []func(*ReplayDescriptor){
		func(d *ReplayDescriptor) { d.SchemaVersion = 1 },
		func(d *ReplayDescriptor) { d.AlgorithmVersion = 2 },
		func(d *ReplayDescriptor) { d.ExecutionVersion = 2 },
		func(d *ReplayDescriptor) { d.MathProfile = "other" },
		func(d *ReplayDescriptor) { d.Algorithm = "unknown" },
	} {
		descriptor := valid
		mutate(&descriptor)
		_, err := ValidateReplayDescriptor(descriptor)
		assertRandomErrorCode(t, err, "UNSUPPORTED_REPLAY_VERSION")
	}
	for _, mutate := range []func(*ReplayDescriptor){
		func(d *ReplayDescriptor) { d.SeedMaterial = "abc" },
		func(d *ReplayDescriptor) { d.SeedMaterial = "0123456789ABCDEF0123456789ABCDEF" },
		func(d *ReplayDescriptor) { d.PlanFingerprint = "bad" },
		func(d *ReplayDescriptor) { d.Origin = "unknown" },
	} {
		descriptor := valid
		mutate(&descriptor)
		_, err := ValidateReplayDescriptor(descriptor)
		assertRandomErrorCode(t, err, "INVALID_REPLAY")
	}
	encoded, _ := json.Marshal(valid)
	var object map[string]any
	_ = json.Unmarshal(encoded, &object)
	object["extra"] = true
	_, err = ValidateReplayDescriptor(object)
	assertRandomErrorCode(t, err, "INVALID_REPLAY")
	extra, _ := json.Marshal(object)
	var decoded ReplayDescriptor
	assertRandomErrorCode(t, json.Unmarshal(extra, &decoded), "INVALID_REPLAY")
	delete(object, "extra")
	delete(object, "origin")
	_, err = ValidateReplayDescriptor(object)
	assertRandomErrorCode(t, err, "INVALID_REPLAY")
	if err := decoded.UnmarshalJSON(append(encoded, []byte(" {}")...)); err == nil {
		t.Fatal("trailing JSON accepted")
	}
}
